lib/resourcemerge/apiext_test.go

```diff
@@ -0,0 +1,190 @@
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: []byte("CA bundle added by the ca operator"),
+							},
+						},
+					},
+				},
+			},
+			required: apiextv1.CustomResourceDefinition{
+				Spec: apiextv1.CustomResourceDefinitionSpec{
+					Conversion: &apiextv1.CustomResourceConversion{
+						Strategy: apiextv1.WebhookConverter,
+						Webhook: &apiextv1.WebhookConversion{
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: nil,
+							},
+						},
+					},
+				},
+			},
+
+			expectedModified: false,
+			expected: apiextv1.CustomResourceDefinition{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						"service.beta.openshift.io/inject-cabundle": "true",
+					},
+				},
+				Spec: apiextv1.CustomResourceDefinitionSpec{
+					Conversion: &apiextv1.CustomResourceConversion{
+						Strategy: apiextv1.WebhookConverter,
+						Webhook: &apiextv1.WebhookConversion{
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: []byte("CA bundle added by the ca operator"),
+							},
+						},
+					},
+				}},
+		},
+		{
+			name: "remove injected caBundle when the annotation `...inject-cabundle=true` is not set",
+			existing: apiextv1.CustomResourceDefinition{
+				Spec: apiextv1.CustomResourceDefinitionSpec{
+					Conversion: &apiextv1.CustomResourceConversion{
+						Strategy: apiextv1.WebhookConverter,
+						Webhook: &apiextv1.WebhookConversion{
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: []byte("CA bundle added by the user"),
+							},
+						},
+					},
+				},
+			},
+			required: apiextv1.CustomResourceDefinition{
+				Spec: apiextv1.CustomResourceDefinitionSpec{
+					Conversion: &apiextv1.CustomResourceConversion{
+						Strategy: apiextv1.WebhookConverter,
+						Webhook: &apiextv1.WebhookConversion{
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: nil,
+							},
+						},
+					},
+				},
+			},
+
+			expectedModified: true,
+			expected: apiextv1.CustomResourceDefinition{
+				Spec: apiextv1.CustomResourceDefinitionSpec{
+					Conversion: &apiextv1.CustomResourceConversion{
+						Strategy: apiextv1.WebhookConverter,
+						Webhook: &apiextv1.WebhookConversion{
+							ClientConfig: &apiextv1.WebhookClientConfig{
+								CABundle: nil,
+							},
+						},
+					},
+				}},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defaultCustomResourceDefinitionV1(&test.existing, test.existing)
+			defaultCustomResourceDefinitionV1(&test.expected, test.expected)
+			modified := pointer.Bool(false)
+			EnsureCustomResourceDefinitionV1(modified, &test.existing, test.required)
+			if *modified != test.expectedModified {
+				t.Errorf("mismatch modified got: %v want: %v", *modified, test.expectedModified)
+			}
+
+			if !equality.Semantic.DeepEqual(test.existing, test.expected) {
+				t.Errorf("mismatch CRD got: %v want: %v", test.existing, test.expected)
```

Comment:
can we log `cmp.Diff(expected, actual)` so the test output shows the difference on failure?

Reply:
Most definitely! We should maybe think about unifying the output of tests where it's appropriate.

Currently, some of the tests have the output for `!equality.Semantic.DeepEqual(test.existing, test.expected)` as:

```go
if !equality.Semantic.DeepEqual(test.existing, test.expected) {
	t.Errorf("mismatch <KIND> got: %v want: %v", test.existing, test.expected)
}
```

and some as:

```go
if !equality.Semantic.DeepEqual(test.existing, test.expected) {
	t.Errorf("unexpected: %s", cmp.Diff(test.expected, test.existing))
}
```

But that's just a minor thought.

Reply:
Yeah... I think that `cmp.Diff` is always better in these compare-output-against-expected unless you have some really specific use-case. Oftentimes you don't even need `equality.Semantic.DeepEqual` and can do `if diff := cmp.Diff(); diff != "" { //print diff }}`. `cmp.Diff` is not *that* old, I imagine some CVO code predates it. I'd consider these as tech dept and would welcome PRs that clean them up.

Reply:
Yes, you are right. Thank you for the helpful comments.